### PR TITLE
Add "/and" option for DSL rules, fix PDF python dependency issues

### PR DESCRIPTION
**Pull request by caleblf**
This update adds a new mode for filters (and added features) in the DSL, written like so:

```
features:
  conjunction-feature/and:
    ^regex1
    ^regex2

filters:
  ConjunctionFilter/and:
    ^regex1
    ^regex2
```

The feature/filter is satisfied only when each regex is matched by at least one message. This is documented in the default PDF DSL text in a comment.

Also, a failure to build a coherent python environment for the PDF distribution is fixed. There were two issues here:

* `pydantic` (imported by some upstream dependencies of the text diff plugin) was updated and no longer works with the (bad) version of typing_extensions pinned by polytracker
* `imageio`'s (dep of the visual diff plugin) default behavior changed in an update

This changeset pins relevant deps and fixes our use of imageio.

Also, grayscale images of PDFs are now properly converted to RGB before doing the visual comparison (we had been getting a dimension mismatch result in such cases). This affects the output of this parser, so it will rerun for all files in live deployments.

### CODE REVIEW

File: faw/pdf-observatory/common/dsl.ts [hunk 1]
```diff
@@ -140,7 +153,7 @@ function generateDslParser() {
         / filters_pattern_check_expr_compare
 
       filters_pattern_check_expr_compare
-        = left:filters_pattern_check_expr_add WS* symb:(">" / ">=" / "<" / "<=" / "==") WS* right:filters_pattern_check_expr_add
+        = left:filters_pattern_check_expr_add WS* symb:(">=" / ">" / "<=" / "<" / "==") WS* right:filters_pattern_check_expr_add
```
Comment (caleblf) [inline faw/pdf-observatory/common/dsl.ts:156]: Unrelated fix: Due to greedy parsing, `<=`/`>=` must be before `<`/`>` to support all alternatives

File: faw/pdf-observatory/common/processor.ts [hunk 2]
```diff
@@ -42,164 +42,215 @@ export function reprocess(
   // Build sets of filter groups
   const filesWithMessages = Object.entries(reprocessPdfGroups.groups);
 
+  function evalCheck(message: string, check: any, identifier: string): Set<number> {
```
Comment (caleblf) [inline faw/pdf-observatory/common/processor.ts:45]: No change to this function's internals; it was just moved outside of the scope below, and `identifier` was made a parameter.

File: faw/pdf-observatory/common/processor.ts [hunk 3]
```diff
@@ -42,164 +42,215 @@ export function reprocess(
...
-    const fileIndices = new Set<number>();
-
     const rules = patterns.map(p => ({
       pat: new RegExp(p.pat, caseInsensitive ? 'i' : undefined),
       check: p.check,
     }));
-    for (const [message, files] of filesWithMessages) {
-      // Do any of our filter's patterns match this message?
-      let filesSubset = new Set<number>();
+    if (!conjunction) {
```
Comment (caleblf) [inline faw/pdf-observatory/common/processor.ts:167]: No change to the logic of this case

File: faw/pdf-observatory/common/processor.ts [hunk 4]
```diff
@@ -42,164 +42,215 @@ export function reprocess(
...
           }
+        }
+        const acceptanceMessage = `'${identifier}' accepted '${message}'`
+        for (const fileIndex of filesSubset) {
+          decisionsByFileIndex.get(fileIndex)!.info.push(acceptanceMessage);
+          fileIndices.add(fileIndex);
+        }
+      }
+      return fileIndices;
+    } else {
```
Comment (caleblf) [inline faw/pdf-observatory/common/processor.ts:200]: `/and` is implemented here

File: faw/pdf-observatory/ui/src/views/HomeView.vue [hunk 5]
```diff
@@ -158,8 +158,8 @@ mixin confusion-matrix
                 v-expansion-panel-content
                   .decision-reasons
                     v-radio-group(v-model="failReasonsSort" row :label="(failReasons.get(decisionAspectSelected) || []).length + ' error messages, sorted by'")
-                      v-radio(value="total" label="number of files rejected")
-                      v-radio(value="unique" label="uniquely rejected")
+                      v-radio(value="total" label="number of files affected")
```
Comment (caleblf) [inline faw/pdf-observatory/ui/src/views/HomeView.vue:161]: Changed this wording due to preference; lmk if you disagree

File: faw/pdf-observatory/ui/src/views/HomeView.vue [hunk 6]
```diff
@@ -1132,12 +1133,30 @@ export default Vue.extend({
         text.push(t);
         text.push('\n');
       };
+      add('features:');
```
Comment (caleblf) [inline faw/pdf-observatory/ui/src/views/HomeView.vue:1136]: Unrelated fix: generation of the extra features was missing

**Comment by wwoods** [comment]
Checking metric inclusion range, there's a bug:

image

Looking through the rest...

**Review by wwoods — APPROVED** [review]
Looks great; a few minor comments but no changes requested.

File: faw/pdf-observatory/common/dsl.ts (hunk 1 above)
Comment (caleblf) [reply faw/pdf-observatory/common/dsl.ts:156]: Apparently pegjs can only backtrack like that when the alternatives are at a higher level. The DSL didn't accept `>=`.

> expression1 / expression2 / ... / expressionn
> Try to match the first expression, if it does not succeed, try the second one, etc. Return the match result of the first successfully matched expression. If no expression matches, consider the match failed.
